index.bs

```diff
@@ -2039,24 +2139,53 @@ input parameters:
 : |allowCredentialDescriptorList|
 :: An optional list of {{PublicKeyCredentialDescriptor}}s describing credentials acceptable to the [=[RP]=] (possibly filtered
     by the client), if any.
+: |requireUserVerification|
+:: The [=effective user verification requirement for assertion=], a Boolean value provided by the client.
 : |extensions|
 :: A [=map=] from [=extension identifiers=] to their [=authenticator extension inputs=], created by the client based on the
     extensions requested by the [=[RP]=], if any.
 
 When this method is invoked, the [=authenticator=] must perform the following procedure:
 1. Check if all the supplied parameters are syntactically well-formed and of the correct length. If not, return an error code
     equivalent to "{{UnknownError}}" and terminate the operation.
+1. Let |isFirstFactorOperation| be a Boolean value, set to `true` if and only if |allowCredentialDescriptorList| was not supplied.
+1. If at least one of |requireUserVerification| and |isFirstFactorOperation| is `true` and the authenticator cannot perform [=user
+    verification=], return an error code equivalent to "{{ConstraintError}}" and terminate the operation.
 1. If |allowCredentialDescriptorList| was not supplied, set it to a list of all credentials stored for |rpId| (as determined by
     an exact match of |rpId|).
-1. Remove any items from |allowCredentialDescriptorList| that do not match a credential bound to this authenticator. A match 
-    occurs if a credential matches <code>|rpId|</code> and an |allowCredentialDescriptorList| item's <code>|allowCredentialDescriptorList| 
-    {{PublicKeyCredentialDescriptor/id}}</code> and <code>|allowCredentialDescriptorList|{{PublicKeyCredentialDescriptor/type}}</code>. 
+1. Remove any items from |allowCredentialDescriptorList| that do not match a credential bound to this authenticator. A match
+    occurs if a credential matches <code>|rpId|</code> and an |allowCredentialDescriptorList| item's
+    {{PublicKeyCredentialDescriptor/id}} and {{PublicKeyCredentialDescriptor/type}} members.
 1. If |allowCredentialDescriptorList| is now empty, return an error code equivalent to "{{NotAllowedError}}" and terminate the
     operation.
+
+1. If |isFirstFactorOperation| is `true`, perform [=user verification=]. If [=user verification=] fails, return an error code
+    equivalent to "{{NotAllowedError}}" and terminate the operation.
+
+    Remove any items from |allowCredentialDescriptorList| that match a credential that was created by a [=user verified|verified
+    user=] other than the current [=user verified|verified user=].
+
 1. Prompt the user to select a [=public key credential|credential=] |selectedCredential| from |allowCredentialDescriptorList|.
```

Comment:
prompting the user to select ought to occur prior to performing user verification, as currently specified, yes?

Reply:
Not for first factor authentications, i.e. when `allowCredentialDescriptorList` is not given. That would mean that an attacker could trivially list the contents of a stolen authenticator.

This does not apply for when `allowCredentialDescriptorList` is given, though, since in that case the attacker already knows the credential ID. The only new information the attacker could gain in that case is the user handle, which is intended to be unlikely to be duplicated between RPs, and it's likely the attacker already knows it if they know the credential ID.